src/Dialect/ONNX/ElementsAttr/ElementsAttrBuilder.cpp

```diff
@@ -885,22 +901,47 @@ ElementsAttr ElementsAttrBuilder::reduce(ElementsAttr elms,
 
   ShapedType reducedType = type.clone(reducedShape);
   return fromWideNums(reducedType, [&](MutableArrayRef<WideNum> dstNums) {
-    // Traverse and populate each element d in dstNums.
-    for (auto &idxoffs : StridesRange<1>(reducedShape, {reducedStrides})) {
-      WideNum &d = dstNums[idxoffs.flattenedIndex];
-      int64_t srcPos = idxoffs[0];
-      // Traverse all the elements that reduce together into d.
-      // srcNums elements may be repeated if there are zeros in axesStrides.
-      StridesRange<1> axesRange(axesShape, {axesStrides});
-      auto axesIter = axesRange.begin();
-      auto axesEnd = axesRange.end();
-      assert(axesIter->at(0) == 0 && "initial src offset must be zero");
-      d = srcNums.get()[srcPos];
-      while (++axesIter != axesEnd) {
-        int64_t srcOffset = axesIter->at(0);
-        d = reducer(d, srcNums.get()[srcPos + srcOffset]);
+    StridesRange<1> sRange(reducedShape, {reducedStrides});
+    SmallVector<std::pair<int64_t, uint64_t>, 4> batch;
+    for (auto &idxoffs : sRange)
+      batch.emplace_back(std::make_pair(idxoffs.flattenedIndex, idxoffs[0]));
+
+    std::mutex mtx;
+    size_t beginOffset = 0;
+    auto fetchBatch = [&](size_t threadNumber) {
+      // Each thread fetches the same batch size. The remainder is set in the
+      // threads with small thread number.
+      const std::lock_guard<std::mutex> lock(mtx);
+      size_t batchSize = batch.size() / ctx->getNumThreads();
+      size_t batchSizeMod = batch.size() % ctx->getNumThreads();
+      if (threadNumber < batchSizeMod)
+        batchSize += 1;
+      auto batchBegin = batch.begin() + beginOffset;
+      auto batchEnd = batchBegin + batchSize;
+      beginOffset += batchSize;
+      return llvm::make_range(batchBegin, batchEnd);
+    };
+
+    auto work = [&](size_t threadNumber) {
+      auto batch = fetchBatch(threadNumber);
+      // Traverse and populate each element d in dstNums.
+      for (auto b : batch) {
+        WideNum &d = dstNums[b.first];
+        int64_t srcPos = b.second;
+        // Traverse all the elements that reduce together into d.
+        // srcNums elements may be repeated if there are zeros in axesStrides.
+        StridesRange<1> axesRange(axesShape, {axesStrides});
+        auto axesIter = axesRange.begin();
+        auto axesEnd = axesRange.end();
+        assert(axesIter->at(0) == 0 && "initial src offset must be zero");
+        d = srcNums.get()[srcPos];
+        while (++axesIter != axesEnd) {
+          int64_t srcOffset = axesIter->at(0);
+          d = reducer(d, srcNums.get()[srcPos + srcOffset]);
+        }
       }
-    }
+    };
+    parallelFor(ctx, 0, ctx->getNumThreads(), work);
   });
 }
```

Comment:
I also assume that the work up there assumes that there are `batch.size()` reductions that can all be done in parallel.

Since we have for quantization "whole tensor" quantization, we have cases where we have only 1 reduction.
That can also be done in parallel. Say you have 1000 elements and 10 threads. Each thread process its own 100 numbers, and save its result in its location in an array of 10 partial sum. Then after the parallel region, just reduce these 10 values sequentially. You will still get a near 10x speedup.

Also, should we check if that if the `batch.size` is small, we may want to do things sequentially? It would probably be good in case we have a few very small tensors. You can easily print out the sizes on stderr for a few benchmarks and see if you have such cases.